cpp/oneapi/dal/detail/communicator.hpp

```diff
@@ -0,0 +1,741 @@
+                        const array<T>& recv,
+                        const std::int64_t root = -1) const {
+#ifdef ONEDAL_ENABLE_ASSERT
+        check_if_same_send_count(send.get_count(), root);
+
+        // Check if recv allocated count is enough to store all the received values
+        // Note: `recv` must be allocated at least on root node
+        const std::int64_t min_recv_count = get_min_recv_count(send.get_count());
+        if (is_root_rank(root)) {
+            ONEDAL_ASSERT(recv.get_count() >= min_recv_count);
+        }
+#endif
+
+        if (send.get_count() == 0) {
+            if (is_root_rank(root)) {
+                ONEDAL_ASSERT(recv.get_count() == 0);
+            }
+            return spmd_request{};
+        }
+
+        T* recv_ptr = nullptr;
+        if (is_root_rank(root)) {
+            ONEDAL_ASSERT(recv.has_mutable_data());
+            recv_ptr = recv.get_mutable_data();
+        }
+
+        spmd_request request;
+
+        __ONEDAL_IF_QUEUE__(send.get_queue(), {
+            auto q = send.get_queue().value();
+
+            if (is_root_rank(root)) {
+                ONEDAL_ASSERT(recv.get_queue().has_value());
+                ONEDAL_ASSERT(recv.get_queue().value() == q);
+            }
+
+            request = gather( //
+                q,
+                send.get_data(),
+                send.get_count(),
+                recv_ptr,
+                send.get_count(),
+                root);
+        });
+
+        __ONEDAL_IF_NO_QUEUE__(send.get_queue(), {
+            request = gather( //
+                send.get_data(),
+                send.get_count(),
+                recv_ptr,
+                send.get_count(),
+                root);
+        });
+
+        return request;
+    }
+
+    /// Collects data from all the ranks within a communicator into a single buffer.
+    /// The data size send by each rank may be different.
+    ///
+    /// @param send_buf   The send buffer
+    /// @param send_count The number of elements of `dtype` in `send_buf`
+    /// @param recv_buf   The receiveing buffer, must contain at least
+    ///                   `rank_count * recv_count` elements,
+    ///                   significant only at `root`
+    /// @param recv_count The number of elements of `dtype` received from
+    ///                   each rank, must contain at least `rank_count` elements,
+    ///                   significant only at `root`
+    /// @param displs     Entry $i$ specifies the displacement relative to
+    ///                   `recv_buf` at which to place the incoming data
+    ///                   from process $i$, must contain at least `rank_count`
+    ///                   elements, significant only at `root`
+    /// @param dtype      The type of elements in the passed buffers
+    /// @param root       The rank of the receiving process
+    ///
+    /// @return The object to track the progress of the operation
+    spmd_request gatherv(const byte_t* send_buf,
```

Comment:
could we add a simplified gatherv later? I mean with recv_counts array but without explicit displacements (calculated from recv_counts

Reply:
No problem it can be added without losing BC